BUG: Ignore mirror caching when no internet

pllim opened the pull request:
Address a bug reported by @parejkoj in #7653 where obtaining cached download fails when there is no internet. This _does not_ address the original feature request in #7653 (i.e., providing a user-defined `sites.json`).

@eteq , can you please review, given you also reviewed #6987 that introduced this bug?

@bsipocz , not sure if this can be cleanly backported to 2.x though #6987 was put in 2.0.4. What should I milestone this to? Do you need a separate PR straight to 2.x branch? Please advise.

Thanks!

parejkoj commented:
Is there any way to write a unittest for this? It looks like there's a `pytest-socket` plugin to disable the internet during tests: https://github.com/miketheman/pytest-socket

pllim commented:
> Is there any way to write a unittest for this? 

@drdavella , is there some kind of magical thingies :unicorn: I can use from `pytest-remotedata`?

pllim commented:
FWIW, I did test this patch on my laptop with internet turned off. :sweat_smile:

drdavella commented:
`pytest-remotedata` should disable internet access for all tests if `--remote-data=none`.

bsipocz commented:
This looks backportable, so I've milestoned as 2.0.10 and will try to remember to double check the behaviour (but having a test doing that for me would be even more awesome)

pllim commented:
Incidentally, the test I refactored back in #7575 was not quite right (it wasn't exactly testing the thing I designed for it to test). So I fixed it here and also incidentally, that test also somehow tests this issue, because `urlopen()` emits `URLError` for local file URI that is a directory, just as if there is no internet. Win, win!

p.s. Given how careless I was at #7575, extra pairs of eyes to check are appreciated!

pllim commented:
p.s. @bsipocz , I have a feeling that this might need a manual backport. Just let me know if that is indeed the case, I'll be happy to open another PR straight to 2.x. Thanks!

bsipocz commented:
Yeah, the diff got much larger. I'll definitely ping you if we get there and if that's a bridge we need to cross.

bsipocz commented:
@eteq - any chance for a review here, so this can land in RC2?

eteq commented on astropy/utils/data.py:
```diff
@@ -1001,9 +1001,11 @@ def download_file(remote_url, cache=False, show_progress=True, timeout=None):
+            with urllib.request.urlopen(conf.dataurl, timeout=timeout) as remote:
+                _dataurls_to_alias[conf.dataurl] = [conf.dataurl, remote.geturl()]
+        except urllib.error.URLError:  # Host unreachable
+            _dataurls_to_alias[conf.dataurl] = [conf.dataurl]
```
```suggestion
           msg = ('url "{}" was requested in download_file, but was not found.'
                        ' Will try cached file if it exists.')
           warn(AstropyWarning(msg.format(conf.dataurl)))
            _dataurls_to_alias[conf.dataurl] = [conf.dataurl]
```

pllim replied:
@eteq , personally, I would prefer *not* emitting a warning. This is because this line was only to cache an alias of the original URL. Hence, user will not find the warning make any sense unless user knows the internal mechanics of this function. If this warning turns out to be necessary, we can always follow up in a future PR. Thank you for your understanding!

bsipocz commented:
Thanks @pllim!

bsipocz commented:
f2f decision was to not backport this to LTS.
